hitl/loopback.py

```diff
@@ -0,0 +1,148 @@
+
+    def __init__(self, client, prefix):
+        """ Constructor. """
+        self.client = client
+        self._telemetry = []
+        self.client.on_message = self.handle_telemetry
+        self._telemetry_topic = f'{prefix}/telemetry'
+        self.client.subscribe(self._telemetry_topic)
+
+
+    def handle_telemetry(self, _client, topic, payload, _qos, _properties):
+        """ Handle incoming telemetry messages over MQTT. """
+        assert topic == self._telemetry_topic
+        self._telemetry.append(json.loads(payload))
+
+
+    def get_latest(self):
+        """ Get the latest telemetry message received. """
+        return self._telemetry[-1]
+
+
+async def test_loopback(miniconf, telemetry, set_point):
+    """ Test loopback operation of Stabilizer.
+
+    Note:
+        Loopback is tested by configuring DACs for static output and verifying telemetry reports the
+        ADCs are measuring those values. Output OUTx should be connected in a loopback configuration
+        to INx on the device.
+
+    Args:
+        miniconf: The miniconf configuration interface.
+        telemetry: a helper utility to read inbound telemetry.
+        set_point: The desired output voltage to test.
+    """
+    print(f'Testing loopback for Vout = {set_point:.2f}')
+    print('---------------------------------')
+    # Configure the IIRs to output at the set point
+    await miniconf.command('iir_ch/0/0', static_iir_output(set_point), retain=False)
+    await miniconf.command('iir_ch/1/0', static_iir_output(set_point), retain=False)
+    await miniconf.command('telemetry_period', 1, retain=False)
+
+    # Wait for telemetry values to update.
+    await asyncio.sleep(5.0)
+
+    # Verify the ADCs are receiving the setpoint voltage.
+    tolerance = max(0.05 * set_point, MINIMUM_VOLTAGE_ERROR)
+    latest_values = telemetry.get_latest()
+    print(f'Latest telemtry: {latest_values}')
+
+    assert abs(latest_values['adcs'][0] - set_point) < tolerance
+    assert abs(latest_values['adcs'][1] - set_point) < tolerance
+    print('PASS')
+    print('')
+
+
+def main():
+    """ Main program entry point. """
+    parser = argparse.ArgumentParser(description='Loopback tests for Stabilizer HITL testing',)
+    parser.add_argument('prefix', type=str,
+                        help='The MQTT topic prefix of the target')
+    parser.add_argument('--broker', '-b', default='mqtt', type=str,
+                        help='The MQTT broker address')
+
+    args = parser.parse_args()
+
+    async def test():
+        """ The actual testing being completed. """
+        interface = await Miniconf.create(args.prefix, args.broker)
+        telemetry = await TelemetryReader.create(args.prefix, args.broker)
+
+        # Test loopback with a static 1V output of the DACs.
+        await test_loopback(interface, telemetry, 1.0)
+
+        # Repeat test with AFE = 2x
+        print('Configuring AFEs to 2x input')
+        await interface.command('afe/0', "G2", retain=False)
```

Comment:
The setup should most of the mqtt settings, not just the non-default ones. And I would inline `test_loopback` as there seems to be no clear distinctions which settings are in there and which are here.